File: server/java/render/src/main/java/com/metaobjects/render/extract/ExtractionResult.java
package com.metaobjects.render.extract;

import java.util.Objects;

/** Typed result of a generated extractLenient(...): best-effort record (null components where lost/malformed) + report. */
public record ExtractionResult<T>(T data, ExtractionReport report) {
    public ExtractionResult {
        Objects.requireNonNull(report, "report");
    }

    /**
     * Strict opt-in gate (Phase B). Returns {@link #data()} when no required field was lost or
     * malformed; otherwise throws a {@link ExtractException} naming the unusable paths.
     *
     * <p>Extract itself never throws — this is the explicit "treat a lost required field as
     * an error" escape hatch for callers who want it.</p>
     *
     * @return {@link #data()} (may itself be {@code null}/partial for non-required losses)
     * @throws ExtractException iff {@code report().hasLostRequired() || report().hasMalformedRequired()}
     */
    public T orThrow() {
        if (report.hasLostRequired() || report.hasMalformedRequired()) {
            throw new ExtractException(report.lostRequired(), report.malformedRequired());
        }
        return data;
    }
}

File: server/csharp/MetaObjects.Render/Extract/ExtractionResult.cs
namespace MetaObjects.Render.Extract;

/// <summary>
/// Typed result of a generated <c>Extract(…)</c> call: best-effort typed record
/// (null-component values where fields were lost/malformed) plus the extraction report.
/// </summary>
public sealed record ExtractionResult<T>(T Data, ExtractionReport Report)
{
    /// <summary>
    /// Strict opt-in gate (Phase B). Returns <see cref="Data"/> when no required field was lost
    /// or malformed; otherwise throws a <see cref="ExtractException"/> naming the unusable paths.
    ///
    /// <para>Extract itself never throws — this is the explicit "treat a lost required field as
    /// an error" escape hatch for callers who want it.</para>
    /// </summary>
    /// <returns><see cref="Data"/> (may itself be null/partial for non-required losses).</returns>
    /// <exception cref="ExtractException">iff <see cref="ExtractionReport.HasLostRequired"/> or
    /// <see cref="ExtractionReport.HasMalformedRequired"/>.</exception>
    public T OrThrow()
    {
        if (Report.HasLostRequired() || Report.HasMalformedRequired())
            throw new ExtractException(Report);
        return Data;
    }
}
